src/main/java/seedu/address/model/reminder/UniqueReminderList.java

```diff
@@ -0,0 +1,139 @@
+package seedu.address.model.reminder;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.LocalDate;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.Model;
+import seedu.address.model.person.Person;
+
+/**
+ * A list of Reminders that enforces uniqueness between its elements and does not allow nulls.
+ * A Reminder is considered unique by comparing using {@code Reminder#equals(Reminder)}. This will allow the reminder
+ * to be added or deleted to maintain the uniqueness of the ReminderList
+ * Supports a minimal set of list operations.
+ *
+ */
+public class UniqueReminderList implements Iterable<Reminder> {
+
+    private boolean isReminderListDirty = false;
+
+    private final Model model;
+    private final ObservableList<Reminder> internalList;
+    private final ObservableList<Reminder> internalUnmodifiableList;
+
+    /**
+     * Constructor of UniqueReminderList
+     */
+    public UniqueReminderList(Model model) {
+        this.model = model;
+        internalList = FXCollections.observableArrayList();
+        internalUnmodifiableList = FXCollections.unmodifiableObservableList(internalList);
+    }
+
+    /**
+     * Sets the reminder list to be dirty.
+     * To be called when the reminder list is modified.
+     */
+    public void setReminderListDirty() {
+        isReminderListDirty = true;
+    }
+
+    /**
+     * Returns true if the list contains no Reminders.
+     */
+    public boolean isEmpty() {
+        return internalList.isEmpty();
+    }
+
+    /**
+     * Returns true if the list contains an equivalent Reminder as the given argument.
+     */
+    public boolean contains(Reminder toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::equals);
+    }
+
+    /**
+     * Returns the backing list as an unmodifiable {@code ObservableList}.
+     */
+    public ObservableList<Reminder> asUnmodifiableObservableList() {
+        return internalUnmodifiableList;
+    }
+
+    /**
+     * Returns the earliest reminder time in the {@code UniqueReminderList}
+     */
+    public long getEarliestReminderTime() {
+        if (internalList.isEmpty()) {
+            return -1;
+        }
+        return internalList.stream().mapToLong(Reminder::getDueTime).min().getAsLong();
+    }
+
+    /**
+     * Returns the list of reminders associated with a specific date.
+     *
+     * @param date The date that which is used to retrieve the list of reminders.
+     * @return the list of reminders mapped from the given date.
+     */
+    public ObservableList<Reminder> getRemindersAfterDate(LocalDate date) {
```

Comment:
Doesn't returning a new `ObservableList` every time defeat the purpose of using `ObservableList`s?

We should have a single instance of `ObservableList` that we add listeners to, and when the internal contents of that list changes, we can then run updates according to the new value.

Reply:
Maybe we can do something similar to how `ModelManager`'s `filteredPersons` works, using a `FilteredList`.

Reply:
This one is @lilozz2's code. Feel free to edit though

Reply:
I will change this when I make the ui for reminders